### DISCUSSION TITLE
how is context block generated?

### DISCUSSION THREAD

**Discussion by dharapvj** [romkatv/powerlevel10k#2731] [General]
I have weird issue where.. if I run powerlevel10k inside a rootless container - I do not get `user@hostname` block displayed. But same container, if I run rootful, I get the required context.

So I am trying to understand how the hostname is read by powerlevel10k.

HOSTNAME env variable is set and I can echo it.

Basically `%n@%m` where are values for `n` and `m` get set?

**Comment by romkatv** [comment]
> Basically `%n@%m` where are values for `n` and `m` get set?

This is a zsh question rather than a p10k one. As far as I recall, the hostname can be overridden with `HOST` but I don't think there is a similar parameter for the username.

```text
% HOST=blah print -P '%m'
blah
```

**Reply by dharapvj** [reply]
First of all - thank you for quick revert and giving me pointer! Much appreciated.

Setting HOST is not making the trick for me. I was already setting HOSTNAME but without much difference.

Somehow.. if I hardcode my state as REMOTE after [this line](https://github.com/romkatv/powerlevel10k/blob/master/internal/p10k.zsh#L1591) like `state="REMOTE"` then my prompt start to work.

That tells me that if the state is DEFAULT, some variable remains unset.. but the for-loop block following this code is like a foreign language for me who hardly knows zsh. Any hints why DEFAULT does not get right values but REMOTE does?

My variables block related to context in `.p10k.zsh` is as follows:

```shell
##################################[ context: user@hostname ]##################################
  # Context color when running with privileges.
  typeset -g POWERLEVEL9K_CONTEXT_ROOT_FOREGROUND=1
  typeset -g POWERLEVEL9K_CONTEXT_ROOT_BACKGROUND=0
  # Context color in SSH without privileges.
  typeset -g POWERLEVEL9K_CONTEXT_{REMOTE,REMOTE_SUDO}_FOREGROUND=3
  typeset -g POWERLEVEL9K_CONTEXT_{REMOTE,REMOTE_SUDO}_BACKGROUND=0
  # Default context color (no privileges, no SSH).
  typeset -g POWERLEVEL9K_CONTEXT_FOREGROUND=3
  typeset -g POWERLEVEL9K_CONTEXT_BACKGROUND=0

  # Context format when running with privileges: user@hostname.
  typeset -g POWERLEVEL9K_CONTEXT_ROOT_TEMPLATE='%n@%m'
  # Context format when in SSH without privileges: user@hostname.
  typeset -g POWERLEVEL9K_CONTEXT_{REMOTE,REMOTE_SUDO}_TEMPLATE='%n@%m'
  # Default context format (no privileges, no SSH): user@hostname.
  typeset -g POWERLEVEL9K_CONTEXT_TEMPLATE='%n@%m'

  # Don't show context unless running with privileges or in SSH.
  # Tip: Remove the next line to always show context.
  typeset -g POWERLEVEL9K_CONTEXT_{DEFAULT,SUDO}_{CONTENT,VISUAL_IDENTIFIER}_EXPANSION=

  # Custom icon.
  # typeset -g POWERLEVEL9K_CONTEXT_VISUAL_IDENTIFIER_EXPANSION='⭐'
  # Custom prefix.
  # typeset -g POWERLEVEL9K_CONTEXT_PREFIX='with '
```

**Comment by romkatv** [comment]
Have you seen this block?

https://github.com/romkatv/powerlevel10k/blob/master/config/p10k-rainbow.zsh#L983-L985

What happens if you heed the tip?

There is also this: https://github.com/romkatv/powerlevel10k?tab=readme-ov-file#how-do-i-add-username-andor-hostname-to-prompt

**Comment by dharapvj** [comment]
aah.. I incorrectly assumed that my `docker exec` session would be interpreted as SSH and so I was ignoring that variable.

Thank you for your support. I commented that particular typeset `POWERLEVEL9K_CONTEXT_{DEFAULT,SUDO}_{CONTENT,VISUAL_IDENTIFIER}_EXPANSION=` and it is working as desired.

**Comment by romkatv** [comment]
You can also set parameters within `~/.p10k.zsh` conditionally depending on whether you are in a docker container or not:

```zsh
if [[ -r /proc/1/cpuset(#qN-.) && $(</proc/1/cpuset) == /docker/* ]]; then
  # in docker
  typeset POWERLEVEL9K_CONTEXT_{DEFAULT,SUDO}_TEMPLATE='docker'
else
  # not in docker
  typeset -g POWERLEVEL9K_CONTEXT_{DEFAULT,SUDO}_{CONTENT,VISUAL_IDENTIFIER}_EXPANSION=
fi
```

**Comment by technomadindie** [comment]
My Powerlevel10k prompt is showing incorrect hostname.
Situation : I connect to LSF in terminal which changes the hostname in the current terminal. But powerlevel10k still shows the old hostname before connecting to LSF. When I run `/bin/hostname` command, it shows the correct and current hostname. What should I add in the prompt to make it show the current hostname from /bin/hostname 

This is my P10K config file for context

```
" ##################################[ context: user@hostname ]##################################
  # Context color when running with privileges.
  typeset -g POWERLEVEL9K_CONTEXT_ROOT_FOREGROUND=178
  # Context color in SSH without privileges.
  typeset -g POWERLEVEL9K_CONTEXT_{REMOTE,REMOTE_SUDO}_FOREGROUND=180
  # Default context color (no privileges, no SSH).
  typeset -g POWERLEVEL9K_CONTEXT_FOREGROUND=180

  # Context format when running with privileges: bold user@hostname.
  typeset -g POWERLEVEL9K_CONTEXT_ROOT_TEMPLATE='%B%n@%m'
  # Context format when in SSH without privileges: user@hostname.
  typeset -g POWERLEVEL9K_CONTEXT_{REMOTE,REMOTE_SUDO}_TEMPLATE='%n@%m'
  # Default context format (no privileges, no SSH): user@hostname.
  typeset -g POWERLEVEL9K_CONTEXT_TEMPLATE='%n@%m'

  # Don't show context unless running with privileges or in SSH.
  # Tip: Remove the next line to always show context.
  #typeset -g POWERLEVEL9K_CONTEXT_{DEFAULT,SUDO}_{CONTENT,VISUAL_IDENTIFIER}_EXPANSION=

  # Custom icon.
  #typeset -g POWERLEVEL9K_CONTEXT_VISUAL_IDENTIFIER_EXPANSION='⭐'
  # Custom prefix.
  typeset -g POWERLEVEL9K_CONTEXT_PREFIX='%fwith '

  ###[ virtualenv: python virtual environment (https://docs.python.org/3/library/venv.html) ]###
  # Python virtual environment color.
  typeset -g POWERLEVEL9K_VIRTUALENV_FOREGROUND=37
  # Don't show Python version next to the virtual environment name.
  typeset -g POWERLEVEL9K_VIRTUALENV_SHOW_PYTHON_VERSION=false
  # If set to "false", won't show virtualenv if pyenv is already shown.
  # If set to "if-different", won't show virtualenv if it's the same as pyenv.
  typeset -g POWERLEVEL9K_VIRTUALENV_SHOW_WITH_PYENV=false
  # Separate environment name from Python version only with a space.
  typeset -g POWERLEVEL9K_VIRTUALENV_{LEFT,RIGHT}_DELIMITER=
  # Custom icon.
  # typeset -g POWERLEVEL9K_VIRTUALENV_VISUAL_IDENTIFIER_EXPANSION='⭐'

"
```

**Comment by romkatv** [comment]
What is the output of `print -P '%m'`? Is it correct or not?

**Reply by technomadindie** [reply]
It is the old one.

**Reply by technomadindie** [reply]
Please help. I want to share my config file with my team. So, I can check the current hostname by writing the command, but I want a  permanent solution.
Also, that why %m is not showing correct hostname

**Reply by romkatv** [reply]
If you believe that the output of `print -P '%m'` is incorrect, consider reporting a bug against zsh. See https://zsh.sourceforge.io/FAQ/zshfaq06.html#l63. Thankfully, you don't need to mention anything about your zsh setup, including the fact you are using powerlevel10k, because `print -P '%m'` is a basic builtin command.

As a workaround, you can implement a custom prompt segment for p10k and use that instead of the provided `context`. See `p10k help segment`. Alternatively, you can set `HOST` when you "connect to LSF in terminal" (quoted because I don't really know what that means), and zsh will respect that.

**Reply by technomadindie** [reply]
Okay Sure, I'll file a bug in zsh. But for now, can you help in what should I write in the command, so that it properly extracts hostname from /bin/hostname.

**Reply by romkatv** [reply]
As I mentioned, you can set `HOST` to override the `%m` expansion. You can try it with `HOST=$(/bin/hostname)`.

**Reply by technomadindie** [reply]
Thanks! It worked!
